File: api/app/routers/feed.py
from fastapi import APIRouter, Depends
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession
from starlette.requests import Request

from app.database import get_session
from app.rate_limit import limiter
from app.services.rss_service import generate_rss_feed

router = APIRouter(prefix="/feed", tags=["feed"])


@router.get("/rss")
@limiter.limit("30/minute")
async def rss_feed(
    request: Request,
    session: AsyncSession = Depends(get_session),
):
    rss_xml = await generate_rss_feed(session)
    return Response(
        content=rss_xml,
        media_type="application/rss+xml; charset=utf-8",
    )

File: api/src/routes/feed.ts
import { Hono } from "hono";
import { describeRoute } from "hono-openapi";
import type { DB } from "../database.js";
import { createRateLimiter } from "../middleware/rate-limit.js";
import { errorResponse } from "../openapi.js";
import { generateRssFeed } from "../services/rss-service.js";
import type { AppEnv } from "../types.js";

export function createFeedRoute(db: DB): Hono<AppEnv> {
  const route = new Hono<AppEnv>();

  route.get(
    "/feed/rss",
    createRateLimiter(30),
    describeRoute({
      tags: ["Feed"],
      summary: "RSS 2.0 feed of the 20 newest articles",
      responses: {
        200: {
          description: "RSS feed",
          content: { "application/rss+xml": { schema: { type: "string" } } },
        },
        503: errorResponse("Feed generation failed"),
      },
    }),
    async (c) => {
      try {
        const rssXml = await generateRssFeed(db);
        return new Response(rssXml, {
          headers: {
            "Content-Type": "application/rss+xml; charset=utf-8",
          },
        });
      } catch (err) {
        c.get("logger").error({ err }, "RSS feed generation failed");
        return c.json({ detail: "Feed generation failed" }, 503);
      }
    },
  );

  return route;
}
